### DISCUSSION TITLE
How to create a theme?

### DISCUSSION THREAD

**Discussion by kecrily** [marp-team/marp#115] [Q&A] [answered]
I want to create a theme, but after I read the documentation, it's still not clear how to create a theme and use it. Can you give me an example of a quick start? I am using marp via vscode plugin.

The documents I read were <https://marpit.marp.app/theme-css>

**Comment by yhatt** [comment, marked as answer]
https://marpit.marp.app/theme-css is only describing about the thinking of theme CSS syntax. To use your own theme actually, you have to register the custom CSS theme in your VS Code workspace: https://github.com/marp-team/marp-vscode#use-custom-theme-css-shield

## Tutorial

1. Create a folder for working and open the folder in VS Code.
2. Put `theme.css` into the created folder. It makes a theme named "your-theme".
   
   ```css
   /* @theme your-theme */
   
   @import 'default';
   
   h1 {
     color: red;
   }
   ```
3. Put `markdown.md` into the created folder.
   
   ```markdown
   ---
   marp: true
   theme: your-theme
   ---
   
   # Hello, world!
   ```
4. Hit <kbd>F1</kbd> and run "Preferences: Open Workspace Settings".
5. Search "Themes" and find out `Markdown > Marp: Themes`.
6. Add a path `theme.css`. Now ready to use "your-theme".
7. Open `markdown.md`, and confirm the root of current workspace is the working folder you've made.
7. Open Markdown preview from the this icon:
   
   ![](https://user-images.githubusercontent.com/3993388/118837273-9b809a80-b8ff-11eb-9b14-e71333ac1393.png)
9. Check whether showing "Hello, world!" with red text.
   
   ![](https://user-images.githubusercontent.com/3993388/118837363-b05d2e00-b8ff-11eb-83c3-b47ad53153d8.png)
10. Congratulations, you've created your first CSS theme!

### Excerice

Try changing `color: red;` to `color: blue;` in `theme.css` and save it.

---

The most part of Marp theme is plain CSS so you have to dive into CSS from here. If you are a beginner, strongly recommend to learn CSS first. https://web.dev/learn/css/

**Reply by mingyuecall** [reply]
Dear yhatt. Thank you for your great work!

The similar issue happen at 5 & 6 steps @calidion . I do not know how to direct a path!

Could you help? Thank you!

**Reply by yhatt** [reply]
These are specs for getting custom theme CSS in Marp for VS Code:

- If your editing Markdown file _belongs to the current workspace_, the path to theme CSS should be **a relative path from the root directory of workspace**.
- If the Markdown file was _placed out of the current workspace_, the path to theme CSS _in global settings_ should be **a relative from its directory**.
- If it's _untitled document_, themes in the local file system would not load.

**Reply by mingyuecall** [reply]
Ok then... I kind of cannot understand...
Is the coding here right? Hope you check but do not waste a lot of your time!

![](https://ming-phd.oss-cn-beijing.aliyuncs.com/Xnip2022-08-16_21-45-19.jpg)

Thank you so much!

**Reply by calidion** [reply]
use this way:
```
{
  "markdown.marp.themes": [
    "./theme.css"
  ]
}
```

**Reply by yhatt** [reply]
@MingChen2020 Your theme is located in `我的研究/PPT/themes/theme.css` and the current workspace of VS Code is `我的研究`. So you have to set the custom theme path as relative from `我的研究`: `./PPT/themes/theme.css`

```json
{
  "markdown.marp.themes": [
    "./PPT/themes/theme.css"
  ]
}
```

**Reply by mingyuecall** [reply]
Thank you so much for your patience!!! It works!!!

I do appreciate this project help write slides more easily!

**Reply by zrxdaly** [reply]
Thanks @yhatt for your answer. 
If i understand correctly, the path should not be set in the user json file right?
also i am a bit confused about how to set the path if md file is outside of theme workspace. 
Could you give a small example of this situation? 
For example: i have a folder that contains several theme.css files: /home/name/Awesome-Marp/themes/*.css
but i normally write md file in other folder: /home/name/Myworkdir/exmaple.md
how do i set the path in this case? Thanks a lot.

**Reply by yhatt** [reply]
> If i understand correctly, the path should not be set in the user json file right?

It's also working if set the local file path(s) to user json. However, Marp team recommends to use the workspace setting, to avoid unnecessary file request across unrelated workspaces.

> also i am a bit confused about how to set the path if md file is outside of theme workspace.
> Could you give a small example of this situation?

Whenever using themes of inside the workspace, _Markdown documents should belong to the same workspace too._  VS Code has been designed as prevented loading external CSS from the out of the workspace.

- https://github.com/microsoft/vscode/issues/45260#issuecomment-371428889
- https://github.com/microsoft/vscode/issues/77290#issuecomment-515729785
- https://github.com/microsoft/vscode/issues/96201
- https://github.com/microsoft/vscode/issues/99280
- https://github.com/microsoft/vscode/issues/101991

If allowed fetching from the outside, it means opening the door of nightmare, such as fetching `/etc/passwd` 🎃

**Comment by calidion** [comment]
the path in 6 should be a full path:  `/home/eric/marp/theme.css`

**Reply by yhatt** [reply]
It's wrong. The path should be a relative path from an opening workspace because Marp for VS Code only can open CSS resources located in the current workspace due to the security reason.

This behavior matches as VS Code setting for the styles of default Markdown preview: microsoft/vscode#45260

**Reply by calidion** [reply]
But direcory path is invalid in my vscode env, while the full path one works. Why?

**Reply by calidion** [reply]
the working vscode settings generated by vscode: 
```json
{
  "markdown.marp.themes": [
    "/home/eric/marp/theme.css"
  ]
}
```

**Reply by yhatt** [reply]
Which version of Marp for VS Code extension are you using? If you were using old version of the extension, recommend to update to the latest stable v1.5.2.

**Reply by calidion** [reply]
using this one now.

```json
{
  "markdown.marp.themes": [
    "./theme.css"
  ]
}
```

**Comment by vmgustavo** [comment]
how can I configure marp to use a custom theme without using vscode?

**Reply by yhatt** [reply]
Marp CLI: https://github.com/marp-team/marp-cli#theme

**Comment by dirsigler** [comment]
what about hosting the theme on a git repository and reference the raw file via its HTTP path?

**Reply by zrxdaly** [reply]
this one works better, thanks

**Comment by GuillaumeDua** [comment]
Thanks for the detailed solution mentioned above.  
Any idea why does expanding macros such as `${workspaceFolder}` does not work ?

```jsonc
    "markdown.marp.themes" : [
        ${workspaceFolder}/theme.css
    ],
```

**Reply by yhatt** [reply]
The specification of `markdown.marp.themes` setting is following `markdown.styles` setting provided by VS Code. `markdown.styles` does not provide variables support such as `${workspaceFolder}` so Marp for VS Code also does not.
https://code.visualstudio.com/docs/languages/markdown#_using-your-own-css

If you want to use the theme located in the workspace folder, just using `"theme.css"` would be work.
